crates/db_views/src/post_view.rs

```diff
@@ -433,10 +476,93 @@ pub struct PostQuery<'a> {
   pub is_profile_view: bool,
   pub page: Option<i64>,
   pub limit: Option<i64>,
+  pub page_after: Option<PaginationCursorData>,
+  pub page_before_or_equal: Option<PaginationCursorData>,
```

Comment:
I'm having trouble understanding what this is about. Why is an upper bound necessary, since we have a limit, and scursor pagination should just work like ` where id > cursor and (more filters) limit X` .

Reply:
I thought so too, but sadly that's not the case. PG can apparently use `WHERE id > cursor LIMIT X` to optimize index fetches from a single part of the index, but not when looking at multiple parts of the index (in this case: multiple communities). I even tried with PG 16 beta. 

You can look at this part of the chart i made for reference:

![image](https://github.com/LemmyNet/lemmy/assets/2303841/b2568232-dbeb-4f10-8b0b-02d4c7d96ad5)

And you can try the standalone SQL script a made to reproduce the problem from this comment (on a clean PG db): https://github.com/LemmyNet/lemmy/issues/2877#issuecomment-1674869898

And you can read this comment where I consider alternatives to this solution: https://github.com/LemmyNet/lemmy/issues/2877#issuecomment-1673597190 (the most interesting one: a CROSS JOIN LATERAL)

Reply:
from what i understand that's the main reason why even the first-page fetches  are sometimes horribly slow on current lemmy btw and the initial reason for this pr (better pagination is just a bonus)